### PR TITLE
New TiledMapPacker output, fixes multiple tileset problem

### PR DESCRIPTION
**Pull request by oddlydrawn**
##### Adds/Changes the following

 * Searches for maps in nested folders, regardless of depth
 * Creates an individual atlas for each map (uses the map name sans extension)
 * Fixes loading/rendering problems with multiple tilesets
 * Removes isBlended checking and addition of blended property
 * Removes unused methods, broken methods, minor cleanup throughout TiledMapPacker class
 * Fixes the problem of not locating tilesets with ../ in the map Note: This has only been fixed with the      new functionality (not --combine-tilsets) The old method of combining tilesets into one atlas still has this problem.
 * Old functionality can be called with the option "--combine-tilesets"
 * Adds two classes to test the options and renders the processed maps
 * Changes usage to: Usage: INPUTDIR [OUTPUTDIR] [--strip-unused] [--combine-tilesets] [-v]
 * Adds a new usage prompt:

        Usage: INPUTDIR [OUTPUTDIR] [--strip-unused] [--combine-tilesets] [-v] 
        Processes a directory of Tiled .tmx maps. Unable to process maps with XML tile layer format.

            --strip-unused 
                 omits all tiles that are not used. Speeds up the processing. Smaller tilesets.           
            --combine-tilesets 
                 instead of creating a tileset for each map, this combines the tilesets into some kind of monster tileset. Has problems with tileset location. Has problems with nested folders. Not recommended. 
             -v 
                 outputs which tiles are stripped and included

##### Example:

Running TiledMapPacker on "./in ./out" with an "in" directory with the following structure:

* in/overworld.tmx
* in/dawn/Examples/Blank.tmx
* in/dawn/Examples/Dungeon.tmx
* in/dawn/Examples/Logo.tmx
* in/dawn/Examples/Mine.tmx
* in/dawn/Examples/Town.tmx
* in/dawn/Examples/Underworld.tmx
* in/world1Maps/dungeon1.tmx
* in/world1Maps/dungeon2.tmx
* in/world2Maps/town1.tmx
* in/world2Maps/town2.tmx

would have the following structure in the "out" directory:

* out/Blank.tmx
* out/dungeon1.tmx
* out/dungeon2.tmx
* out/Dungeon.tmx
* out/Logo.tmx
* out/Mine.tmx
* out/overworld.tmx
* out/town1.tmx
* out/town2.tmx
* out/Town.tmx
* out/Underworld.tmx
* out/tileset/Blank.atlas
* out/tileset/dungeon1.atlas
* out/tileset/dungeon1.png
* out/tileset/dungeon2.atlas
* out/tileset/dungeon2.png
* out/tileset/Dungeon.atlas
* out/tileset/Dungeon.png
* out/tileset/Logo.atlas
* out/tileset/Logo.png
* out/tileset/Mine.atlas
* out/tileset/Mine.png
* out/tileset/overworld.atlas
* out/tileset/overworld.png
* out/tileset/town1.atlas
* out/tileset/town2.atlas
* out/tileset/town2.png
* out/tileset/Town.atlas
* out/tileset/Town.png
* out/tileset/Underworld.atlas
* out/tileset/Underworld.png

(Note: maps without a tileset .png were completely empty maps)

Old functionality (now --combine-tilesets) can't handle the nested directories or parent (../) paths for tilesets but if it could run on the same in/out directories it would have the following output:

 * out/{same maps here}
 * out/tileset/packed.atlas
 * out/tileset/packed.png

With all of the tiles from every map inside the packed.png.

##### TL;DR

Changes TiledMapPacker to create one texture atlas per map, adds two classes to
test that I haven't messed things up more than they were.. I think. Fixes multiple tileset loading/rendering problem. Removes isBlended checking, property addition.
 * modified:   CHANGES
 * modified:   extensions/gdx-tools/src/com/badlogic/gdx/tiledmappacker/TiledMapPacker.java
 * new file:   extensions/gdx-tools/src/com/badlogic/gdx/tiledmappacker/TiledMapPackerTest.java
 * new file:   extensions/gdx-tools/src/com/badlogic/gdx/tiledmappacker/TiledMapPackerTestRender.java
 * modified: gdx/src/com/badlogic/gdx/maps/tiled/AtlasTmxMapLoader.java

### CODE REVIEW

**Comment by MobiDevelop** [comment]
Please don't create a new PR for every change. If you make new commits they will automatically be added to the existing PR.

**Comment by oddlydrawn** [comment]
Sorry.

**Comment by MobiDevelop** [comment]
Is that change in AtlasTmxMapLoader compatible with existing stuff?

**Comment by oddlydrawn** [comment]
I asked in IRC if AtlasTmxMapLoader was used for anything other than TiledMapPacker's processed maps but there wasn't anyone that knew with absolute certainty. From my understanding, AtlasTmxMapLoader is only used to process maps with an 'atlas' property and TiledMapPacker is the only thing that adds an 'atlas' property to tiled maps. I could try asking at another time.

**Comment by MobiDevelop** [comment]
People do use it without using the packer.

**Comment by oddlydrawn** [comment]
The call hierarchy in libGDX says that the only files that use AtlasTmxMapLoader are
TiledMapAtlasAssetManagerTest
TiledMapAtlasDirectLoaderTest
Although the asset manager test is the only one that shows up in my LwjglTestStarter.

They both point to data/maps/tiled-atlas-processed/test.tmx
and that tiled map has the tell-tale sign of being processed with tiledMapPacker since the first line has

    property name="atlas" value="tileset/packed.atlas" property name="blended tiles" value=""

If you run TiledMapAtlasAssetManagerTest, there will be artifacts on my branch since it was created with the old TiledMapPacker. If you run the TiledMapPacker tool with tiled-atlas-src as the input and tiled-atlas-processed as the output, they will render better than master. [This is what master's TiledMapAtlasAssetManagerTest looks like](http://i.imgur.com/FVYOq6E.png) and [this is what it renders like in my branch](http://i.imgur.com/qG5o5aG.png). 

The more tiles you have, the more obvious the problems with multiple tilesets are. [Here's a two layers, each with an entire tileset on it](http://i.imgur.com/DLhJJqJ.png) the one on the right is how it currently is in master. The center one has been processed with all the changes in my branch.

I've added the reprocessed maps to my branch.

I'd like to know what else uses AtlasTmxMapLoader to create tests to check for conflicts.

**Comment by MobiDevelop** [comment]
It is a public class... part of the public api. Anybody can use it in their
project. Just because libgdx itself doesn't use it, doesn't mean it isn't
used.

**Comment by oddlydrawn** [comment]
I know. However, the current implementation is broken. AtlasTmxMapLoader is currently expecting for the tileset index for each tileset to begin at zero and end at number of tiles in each tilset then the firstgid for the tileset would be added to that. Since it iterates through every texture region, there would be conflicts with almost every single tile. 

The way I changed it, it now expects that the index of every tile is exactly the index that is used in the Tiled Map Editor. So every tile can now have a unique id, as it should.

**Comment by MobiDevelop** [comment]
As far as I know, the way it works now is that each tileset should be
packed in the atlas by the tileset name, with indices by tile. Whether the
packer packs them that way, I don't know.

**Comment by oddlydrawn** [comment]
It does look like that was the way AtlasTmxMapLoader was written. The packer has been packing every texture region with "packed" as the name, not the tileset's name. There hasn't been a change in the tiled-atlas-processed directory for 2+ years. I'm not sure about the packer since it used to be tiled-preprocessor...? So the AtlasTmxMapLoader rewrite might have broken the packer, even though it seems to have been written for the packer.

I still think having texture region indexes start at zero for every tileset is a mistake. Trying to debug the problem as it is was horrible enough. It seems much nicer to be able to make a map in Tiled, save the tile layer as CSV, look at the index for a particular tile, be able to look up that index in the atlas easily, and know that the index will line up inside the map's data structure. Kind of like having the same variable for one thing throughout a project, you see it somewhere, and you know what it is.

**Comment by MobiDevelop** [comment]
AtlasTmxMapLoader really has nothing to do with the packer. When I rewrote
the maps api, I deleted the packer and someone else revived it.

I would prefer we don't change the behavior of the loader as it is part of
the public api and there isn't a compelling reason to change it.

**Comment by oddlydrawn** [comment]
I still think the current AtlasTmxMapLoader implementation is broken since it doesn't do any kind of checking against tileset name to texture region name. There would still be conflicts with indexes when using more than one tileset.

**Comment by oddlydrawn** [comment]
I'll rewrite the packer and test it. So, AtlasTmxMapLoader should be okay with one atlas per map, a region name equal to the tileset name, with indexes from zero to last tile, right?

**Comment by oddlydrawn** [comment]
To get AtlasTmxMapLoader to load an atlas in line with the intent it was written, where it would load regions according to the tileset name/index as I understand it, [this line](https://github.com/libgdx/libgdx/blob/master/gdx/src/com/badlogic/gdx/maps/tiled/AtlasTmxMapLoader.java#L351) in the loader would need to change to

    String regionsName = name;

Since it gets the layer's name [on this line](https://github.com/libgdx/libgdx/blob/master/gdx/src/com/badlogic/gdx/maps/tiled/AtlasTmxMapLoader.java#L284). Once you do that, it works like you said it should. 

Should I create a bug report with the tests, separate PR, or incorporate that into this PR?

**Comment by oddlydrawn** [comment]
I incorporated the change into this PR. I updated the test files for AtlasTmxMapLoader so they're in line to how you described it should work. The packer now outputs atlas regions formatted in the way you described. With the loader fix, and the updated processed maps, TiledMapAtlasAssetManagerTest is now rendering correctly.

Fixing AtlasTmxMapLoader fixed all of the problems TiledMapPacker was having with multiple layers, at least the problems I am aware of and actively testing for. Now there are no issues in the packer that I know of, except that tile layer format can't be XML since it gets the limitation from TmxMapLoader, from what I understand.

This branch is finished, pending review. Although I'm not sure if I should delete gdx-tools' TiledMapPackerTest and TiledMapPackerTestRender, I really only included them to verify that I didn't mess things up more than they were.

**Comment by MobiDevelop** [comment]
Alrighty, will take some time to fully review due to work and family obligations, but I'll do so soon.
